lib/membrane_live/event_service.ex

```diff
@@ -0,0 +1,208 @@
+      :undefined ->
+        {:ok, pid} = Room.start(event_id, name: {:global, event_id})
+        state = put_in(state, [:pid_to_id, pid], event_id)
+        Process.monitor(pid)
+        {:reply, pid, state}
+
+      pid ->
+        {:reply, {:already_started, pid}, state}
+    end
+  end
+
+  @impl true
+  def handle_info({:event, action, event_id}, state) do
+    previous_users_number = get_in(state.events, [event_id, :users_number]) || 0
+    timer_ref = get_in(state.events, [event_id, :timer_ref])
+
+    users_number =
+      case action do
+        :join -> previous_users_number + 1
+        :leave -> previous_users_number - 1
+      end
+
+    timer_ref =
+      case users_number do
+        0 ->
+          cancel_action(timer_ref)
+          kill_action(event_id)
+
+        1 ->
+          cancel_action(timer_ref)
+          notify_action(event_id)
+
+        2 ->
+          cancel_action(timer_ref)
+          nil
+
+        _other ->
+          nil
+      end
+
+    event = %{users_number: users_number, timer_ref: timer_ref}
+    {:noreply, put_in(state, [:events, event_id], event)}
+  end
+
+  @impl true
+  def handle_info({:notify, event_id}, state) do
+    MembraneLiveWeb.Endpoint.broadcast!("event:" <> event_id, "last_viewer_active", %{
+      timeout: @response_timeout
+    })
+
+    timer_ref = kill_action(event_id)
+    {:noreply, put_in(state, [:events, event_id, :timer_ref], timer_ref)}
+  end
+
+  @impl true
+  def handle_info({:response, event_id, :leave}, state) do
+    __MODULE__.send_kill(event_id)
+    {:noreply, state}
+  end
+
+  @impl true
+  def handle_info({:response, event_id, :stay}, state) do
+    users_number = get_in(state.events, [event_id, :users_number])
+    timer_ref = get_in(state.events, [event_id, :timer_ref])
+
+    cancel_action(timer_ref)
+
+    timer_ref =
+      case users_number do
+        0 -> kill_action(event_id)
+        1 -> notify_action(event_id)
+        _other -> nil
+      end
+
+    {:noreply, put_in(state, [:events, event_id, :timer_ref], timer_ref)}
+  end
+
+  @impl true
+  def handle_info({:kill, event_id}, state) do
+    pid = :global.whereis_name(event_id)
+
+    {event, state} = pop_in(state, [:events, event_id])
+    {_event_id, state} = pop_in(state, [:pid_to_id, pid])
+
+    unless pid == :undefined do
+      Room.kill(pid)
+    end
+
+    unless is_nil(event) do
+      MembraneLiveWeb.Endpoint.broadcast!("event:" <> event_id, "finish_event", %{})
+      MembraneLive.Webinars.mark_webinar_as_finished(event_id)
+    end
+
+    {:noreply, state}
+  end
+
+  @impl true
+  def handle_info({:DOWN, _ref, :process, pid, _reason}, state) do
+    {event_id, state} = pop_in(state, [:pid_to_id, pid])
+    {_event, state} = pop_in(state, [:events, event_id])
+
+    unless is_nil(event_id) do
+      MembraneLiveWeb.Endpoint.broadcast!("event:" <> event_id, "finish_event", %{})
+      MembraneLive.Webinars.mark_webinar_as_finished(event_id)
+    end
```

Comment:
This part is the same as in the handler above. Maybe we could move it to a function?